device/iot-device-samples/custom-sas-token-provider-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/CustomSasTokenProviderSample.java

```diff
@@ -0,0 +1,371 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+package samples.com.microsoft.azure.sdk.iot;
+
+import com.microsoft.azure.sdk.iot.device.*;
+import com.microsoft.azure.sdk.iot.device.transport.IotHubConnectionStatus;
+import org.apache.commons.codec.binary.Base64;
+
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.List;
+
+
+/**
+ * This sample demonstrates how to configure your device client to use a custom SAS token provider instead of
+ * directly providing it the device's symmetric key.
+ */
+public class CustomSasTokenProviderSample
+{
+    private static final int D2C_MESSAGE_TIMEOUT = 2000; // 2 seconds
+    private static List<String> failedMessageListOnClose = new ArrayList(); // List of messages that failed on close
+
+    /**
+     * Helper class for turning symmetric keys into SAS tokens. It also provides some helpful functions around
+     * if this token should be renewed.
+     */
+    protected static class SasToken
```

Comment:
I think it's important to make the distinction that this is a helper class by including that in the name. 
`SampleSasTokenHelper `? or something of that nature.

Reply:
This class is functional enough to replace the similar classes we have defined in our SDK, so I kind of like the name as is.

Reply:
then maybe we should include this as the default implementation of our own that ships along with the interface?

Reply:
But it isn't a SasToken, so the name is misleading, don't you think?

Reply:
Sure then, I can rename this to ```SasTokenHelper```